File: Introduction to Robotics Course/Chapter19/Lesson6/Chapter19_Lesson6.py
import pandas as pd
import numpy as np

# Load CSV exported from simulator or robot logger
df = pd.read_csv("robot_logs.csv")

# Compute error signal
df["e"] = df["y_ref"] - df["y_meas"]

# Group by trial and compute RMS error
def rms(x):
    return np.sqrt(np.mean(np.square(x)))

rms_per_trial = df.groupby("trial_id")["e"].apply(rms)

J_values = rms_per_trial.values
N = len(J_values)

J_mean = np.mean(J_values)
J_std = np.std(J_values, ddof=1)

print("RMS error per trial:")
print(rms_per_trial)
print(f"Mean RMS error: {J_mean:.4f}")
print(f"Std of RMS error: {J_std:.4f}")

# 95% confidence interval for the mean (approximate, using normal quantile)
alpha = 0.05
z = 1.96  # approximate for large N
se = J_std / np.sqrt(N)
print(f"95% CI for mean RMS error: [{J_mean - z * se:.4f}, {J_mean + z * se:.4f}]")
      

File: Introduction_to_Robotics_Course/Chapter19/Lesson6/Chapter19_Lesson6.m
% Load logs exported from the experiment (e.g., from Simulink Data Inspector)
data = readtable("robot_logs.csv");

% Error signal
data.e = data.y_ref - data.y_meas;

trial_ids = unique(data.trial_id);
J_values = zeros(numel(trial_ids), 1);

for i = 1:numel(trial_ids)
    tid = trial_ids(i);
    idx = data.trial_id == tid;
    e = data.e(idx);
    J_values(i) = sqrt(mean(e.^2));
    fprintf("Trial %d RMS error = %.4f\n", tid, J_values(i));
end

J_mean = mean(J_values);
J_std  = std(J_values, 1);   % population std; for sample use std(J_values, 0)

fprintf("Mean RMS error = %.4f\n", J_mean);
fprintf("Std of RMS error = %.4f\n", J_std);

% Example: if J_values come from a Simulink model run multiple times
% for different seeds, this script can be placed in a Live Script that
% also embeds plots and report text.
      
